### PR TITLE
Privacy

### PR DESCRIPTION
**Pull request by paul-szczepanek-arm**

### CODE REVIEW

File: BLE_Privacy/mbed-os.lib [hunk 1]
```diff
@@ -0,0 +1 @@
+https://github.com/ARMmbed/mbed-os/#f9ee4e849f8cbd64f1ec5fdd4ad256585a208360
```
Comment (donatieng) [inline BLE_Privacy/mbed-os.lib]: That's mbed OS 5.8.2 (does that work?), please make it point to mbed OS 5.9 RC1 :) (`6817dc43f9f5216cbe077059fc561fa89a766dc9 `)

File: BLE_Privacy/source/main.cpp [hunk 2]
```diff
@@ -0,0 +1,575 @@
...
+            _whitelist_generated = true;
+
+            /* set the newly created whitelist at the link layer,
+             * see BLUETOOTH SPECIFICATION Version 5.0 | Vol 6, Part B - 4.3 */
+            _ble.gap().setWhitelist(*whitelist);
+        } else {
+            printf("Whitelist failed to generate.\r\n");
+        }
+
+        /* this callback transfer memory ownership to us so we have to dispose of it */
```
Comment (donatieng) [inline BLE_Privacy/source/main.cpp]: Actually we have ownership of it during the whole process.

File: BLE_Privacy/source/main.cpp [hunk 3]
```diff
@@ -0,0 +1,575 @@
...
+
+        _ble.gap().setPeripheralPrivacyConfiguration(&privacy_configuration);
+        _ble.gap().setAdvertisingPolicyMode(Gap::ADV_POLICY_IGNORE_WHITELIST);
+
+        if (!start_advertising()) {
+            return;
+        }
+    };
+
+    /** advertise and filer based on known devices */
```
Comment (donatieng) [inline BLE_Privacy/source/main.cpp]: `filer` --> `filter`

File: BLE_Privacy/source/main.cpp (hunk 2 above)
Comment (paul-szczepanek-arm) [reply BLE_Privacy/source/main.cpp]: we are not allowed to delete it before the callback happens and we do not store a pointer to it so effectively we are handing it over

File: BLE_Privacy/mbed-os.lib (hunk 1 above)
Comment (paul-szczepanek-arm) [reply BLE_Privacy/mbed-os.lib]: yes, thank you

File: BLE_Privacy/source/main.cpp [hunk 4]
```diff
@@ -0,0 +1,578 @@
...
+        }
+
+        /* parse the advertising payload, looking for a discoverable device */
+        for (uint8_t i = 0; i < params->advertisingDataLen; ++i) {
+            /* The advertising payload is a collection of key/value records where
+             * byte 0: length of the record excluding this byte
+             * byte 1: The key, it is the type of the data
+             * byte [2..N] The value. N is equal to byte0 - 1 */
+
+            const uint8_t record_length = params->advertisingData[i];
```
[the remark this answers is not in the document]
Comment (paul-szczepanek-arm) [reply BLE_Privacy/source/main.cpp]: ...in which case I should also not trust that the string is well formatted.

What we really need is the on_scan callback to return a typed list of items and not having the user look at bytes.

File: BLE_Privacy/source/main.cpp (hunk 2 above)
Comment (paul-szczepanek-arm) [reply BLE_Privacy/source/main.cpp]: I will now seemingly backtrack on what I said: lifetime is decided by the user, the whitelist could be statically allocated, in my example I relinquish control but the doc doesn't require that: the only requirement is that you do not access the whitelist when callback is pending.

**Comment by paul-szczepanek-arm** [comment]
addressed your comments

File: BLE_Privacy/source/main.cpp [hunk 5]
```diff
@@ -0,0 +1,582 @@
...
+
+    /* helper functions */
+
+private:
+    bool set_advertising_data()
+    {
+        GapAdvertisingData advertising_data;
+
+        /* add advertising flags */
+        advertising_data.addFlags(GapAdvertisingData::LE_GENERAL_DISCOVERABLE
```
Comment (pan-) [inline BLE_Privacy/source/main.cpp]: If you intend to use the whitelist then remove the discoverable flag.

File: BLE_Privacy/source/main.cpp [hunk 6]
```diff
@@ -0,0 +1,582 @@
...
+             * byte [2..N] The value. N is equal to byte0 - 1 */
+            const uint8_t max_record_length = params->advertisingDataLen - i;
+            const uint8_t record_length = std::min(params->advertisingData[i],
+                                                   max_record_length);
+            const uint8_t type = params->advertisingData[i + 1];
+            const uint8_t *value = params->advertisingData + i + 2;
+
+            if (record_length < 2) {
+                /* malformed record */
+            } else if ((type == GapAdvertisingData::FLAGS)) {
```
Comment (pan-) [inline BLE_Privacy/source/main.cpp]: This bit must be removed if the flag is not set in advertising data.

File: BLE_Privacy/source/main.cpp [hunk 7]
```diff
@@ -0,0 +1,582 @@
...
+        }
+    };
+
+    /** Start BLE interface initialisation */
+    void run()
+    {
+        ble_error_t error;
+
+        /* to show we're running we'll blink every 500ms */
+        _event_queue.call_every(500, this, &PrivacyDevice::blink);
```
[the remark this answers is not in the document]
Comment (paul-szczepanek-arm) [reply BLE_Privacy/source/main.cpp]: queue now is recreated

File: BLE_Privacy/source/main.cpp [hunk 8]
```diff
@@ -0,0 +1,582 @@
...
+
+        if (event->error) {
+            printf("Error during the initialisation\r\n");
+            _event_queue.break_dispatch();
+            return;
+        }
+
+        /* for use by tools we print out own address and also use it
+         * to seed RNG as the address is unique */
+        print_local_address();
```
[the remark this answers is not in the document]
Comment (paul-szczepanek-arm) [reply BLE_Privacy/source/main.cpp]: we use it in our tools and I use it as seed and it shows the identity address which we will later see on the other side

**Comment by paul-szczepanek-arm** [comment]
This stil requires fixing gap so that restarting scanning works

File: BLE_Privacy/source/main.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (pan-) [reply BLE_Privacy/source/main.cpp]: We use it in our tools to identify the device advertising unfortunately this won't work this time as the address is private. I would suggest to add a random vendor specific field in the advertising payload and print it so we can identify the device advertising.

**Comment by paul-szczepanek-arm** [comment]
thanks, fixed

File: BLE_Privacy/source/main.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (paul-szczepanek-arm) [reply BLE_Privacy/source/main.cpp]: It is harmless and it does show the identity address before the privacy is enabled so you can see that you're connecting to the right device - connections is based on name

**Comment by paul-szczepanek-arm** [comment]
thank you

File: BLE_Privacy/source/main.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (donatieng) [reply BLE_Privacy/source/main.cpp]: That's different: Badly formatted buffer: we get garbage that we can ignore; String with length flag set to an invalid value (which is a specific case of the former): undefined behaviour.

Agreed that in the long run which should do that parsing ourselves.


Sure - in the mean time we still need to add that check, as we don't

File: BLE_Privacy/source/main.cpp [hunk 9]
```diff
@@ -0,0 +1,578 @@
...
+            const uint8_t *value = params->advertisingData + i + 2;
+
+            if ((type == GapAdvertisingData::FLAGS)) {
+                /* connect to discoverable devices only */
+                if (!(*value & GapAdvertisingData::LE_GENERAL_DISCOVERABLE)) {
+                    return;
+                }
+            } else if (type == GapAdvertisingData::COMPLETE_LOCAL_NAME) {
+                /* connect based on the name of the device */
+                if (strcmp((const char*)DEVICE_NAME, (const char*)value) == 0) {
```
Comment (donatieng) [inline BLE_Privacy/source/main.cpp]: Actually, there's no guarantee that value is 0-terminated right?

**Pull request merged**